core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocatePrimaryAllocationCommand.java

```diff
@@ -0,0 +1,297 @@
+            }
+
+            if (allocationId != null) {
+                if (allocationId.equals(nodeShardState.allocationId()) == false) {
+                    return getFailedRerouteExplanation(explain, allocation,
+                        "found shard data for shard " + shardId + " on node " + node + " but does not match allocation id (expected ["
+                            + allocation + "] but was [" + nodeShardState.allocationId() + "])");
+                }
+            }
+        } else {
+            if (allocationId != null) {
+                // search for shard with allocation id on all nodes
+                for (Map.Entry<DiscoveryNode, TransportNodesListGatewayStartedShards.NodeGatewayStartedShards> entry : fetchResult.getData().entrySet()) {
+                    if (entry.getValue().storeException() == null && allocationId.equals(entry.getValue().allocationId())) {
+                        discoNode = entry.getKey();
+                        break;
+                    }
+                }
+                if (discoNode == null) {
+                    return getFailedRerouteExplanation(explain, allocation,
+                        "found no shard data for shard " + shardId + " matching allocation id [" + allocationId + "]");
+                }
+            }
+        }
+
+        assert discoNode != null;
+
+        // OK, now we go ahead and allocate
+        RoutingNode routingNode = routingNodes.node(discoNode.id());
+        if (routingNode == null) {
+            if (!discoNode.dataNode()) {
+                return getFailedRerouteExplanation(explain, allocation,
+                    "Allocation can only be done on data nodes, not [" + node + "]");
+            } else {
+                return getFailedRerouteExplanation(explain, allocation,
+                    "Could not find [" + node + "] among the routing nodes");
+            }
+        }
+
+        Decision decision = allocation.deciders().canAllocate(shardRouting, routingNode, allocation);
+        if (ignoreDeciders == false && decision.type() == Decision.Type.NO) {
+            return getFailedRerouteExplanation(explain, allocation,
+                "allocation of " + shardId + " on node " + discoNode + " is not allowed, reason: " + decision);
+        }
+        // go over and remove it from the unassigned
+        for (RoutingNodes.UnassignedShards.UnassignedIterator it = routingNodes.unassigned().iterator(); it.hasNext(); ) {
+            ShardRouting unassigned = it.next();
+            if (unassigned != shardRouting) {
+                continue;
+            }
+            it.initialize(routingNode.nodeId(), unassigned.version(), allocation.clusterInfo().getShardSize(unassigned, ShardRouting.UNAVAILABLE_EXPECTED_SHARD_SIZE));
+            break;
+        }
+        if (ignoreDeciders && decision.type() == Decision.Type.NO) {
+            decision = allocation.decision(Decision.YES, RECOVER_PRIMARY_ALLOCATION_COMMAND, "override deciders");
```

Comment:
This seems mighty dangerous but I get it. I suspect things will start relocating as soon as you do this too.